app/lib/search/search_service.dart

```diff
@@ -444,33 +455,52 @@ class TagsPredicate {
         tag = tag.substring(1);
       }
       if (required) {
-        requiredTags.add(tag);
+        p._values[tag] = true;
       } else {
-        negatedTags.add(tag);
+        p._values[tag] = false;
       }
     }
-    return TagsPredicate(requiredTags: requiredTags, negatedTags: negatedTags);
+    return p;
+  }
+
+  /// Appends [requiredTags] and [prohibitedTags] to the current set of tags,
+  /// and returns a new [TagsPredicate] instance.
+  /// The appended flags take precedence over the current ones.
+  TagsPredicate append(
+      {List<String> requiredTags, List<String> prohibitedTags}) {
+    final p = TagsPredicate();
+    p._values.addAll(_values);
+    requiredTags?.forEach((tag) => _values[tag] = true);
+    prohibitedTags?.forEach((tag) => _values[tag] = false);
+    return p;
+  }
+
+  /// Appends [other] predicate to the current set of tags, and returns a new
+  /// [TagsPredicate] instance.
+  /// The appended flags take precedence over the current ones.
```

Comment:
```
//// If there are conflicting tag predicates, the [other] takes precedence over this [TagsPredicate].
```

We could also consider calling it `merge` or `combine`...

Reply:
`merge` implies in-place update to me, I'm undecided about `combine`.